File: app/src/main/java/net/squanchy/schedule/view/EventsAdapter.java
package net.squanchy.schedule.view;

import android.content.Context;
import android.support.annotation.IntDef;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import java.util.Collections;
import java.util.List;

import net.squanchy.R;
import net.squanchy.schedule.domain.view.Event;

class EventsAdapter extends RecyclerView.Adapter<EventViewHolder> {

    private final LayoutInflater layoutInflater;

    @IntDef(value = {ItemViewType.TYPE_TALK, ItemViewType.TYPE_OTHER})
    @interface ItemViewType {

        int TYPE_TALK = 1;
        int TYPE_OTHER = 2;
    }

    private List<Event> events = Collections.emptyList();

    @Nullable
    private ScheduleViewPagerAdapter.OnEventClickedListener listener;

    EventsAdapter(Context context) {
        setHasStableIds(true);
        layoutInflater = LayoutInflater.from(context);
    }

    @Override
    public long getItemId(int position) {
        return events.get(position).getNumericId();
    }

    void updateWith(List<Event> events, ScheduleViewPagerAdapter.OnEventClickedListener listener) {
        this.events = events;
        this.listener = listener;
    }

    @Override
    @ItemViewType
    public int getItemViewType(int position) {
        Event.Type itemType = events.get(position).getType();
        switch (itemType) {
            case KEYNOTE:
            case TALK:
                return ItemViewType.TYPE_TALK;
            case COFFEE_BREAK:
            case LUNCH:
            case OTHER:
            case REGISTRATION:
            case SOCIAL:
                return ItemViewType.TYPE_OTHER;
            default:
                throw new IllegalArgumentException("Item of type " + itemType + " is not supported");
        }
    }

    @Override
    public EventViewHolder onCreateViewHolder(ViewGroup parent, @ItemViewType int viewType) {
        EventItemView itemView;
        if (viewType == ItemViewType.TYPE_TALK) {
            itemView = (EventItemView) layoutInflater.inflate(R.layout.item_schedule_event_talk, parent, false);
        } else if (viewType == ItemViewType.TYPE_OTHER) {
            itemView = (EventItemView) layoutInflater.inflate(R.layout.item_schedule_event_other, parent, false);
        } else {
            throw new IllegalArgumentException("View type not supported: " + viewType);
        }
        return new EventViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(EventViewHolder holder, int position) {
        holder.updateWith(events.get(position), listener);
    }

    @Override
    public int getItemCount() {
        return events.size();
    }

    public List<Event> events() {
        return events;
    }
}

File: app/src/main/java/net/squanchy/schedule/view/EventsAdapter.kt
package net.squanchy.schedule.view

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import net.squanchy.R
import net.squanchy.schedule.domain.view.Event

internal class EventsAdapter(context: Context) : ListAdapter<Event, EventViewHolder>(DiffCallback) {

    init {
        setHasStableIds(true)
    }

    private val layoutInflater: LayoutInflater = LayoutInflater.from(context)
    private var showRoom = false
    private var showFavorites = false
    private var eventClickListener: OnEventClickListener? = null

    fun updateWith(list: List<Event>, showRoom: Boolean, showFavorites: Boolean, listener: OnEventClickListener) {
        this.showRoom = showRoom
        this.eventClickListener = listener

        if (this.showFavorites != showFavorites) {
            // We want to refresh the favourite visualization too which is not part of the model and thus
            // would get ignored by the DiffCallback when comparing things. We go nuclear and trigger
            // a full dataset change notification.
            this.showFavorites = showFavorites
            notifyDataSetChanged()
        }
        super.submitList(list)
    }

    override fun getItemId(position: Int) = getItem(position).numericId

    override fun getItemViewType(position: Int): Int {
        val itemType = getItem(position).type
        return when (itemType) {
            Event.Type.KEYNOTE,
            Event.Type.TALK,
            Event.Type.WORKSHOP -> ItemViewType.TYPE_TALK.ordinal

            Event.Type.COFFEE_BREAK,
            Event.Type.LUNCH,
            Event.Type.OTHER,
            Event.Type.REGISTRATION,
            Event.Type.SOCIAL -> ItemViewType.TYPE_OTHER.ordinal
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): EventViewHolder {
        val itemViewType = ItemViewType.values()[viewType]
        val itemView: EventItemView = when (itemViewType) {
            ItemViewType.TYPE_TALK -> layoutInflater.inflate(R.layout.item_schedule_event_talk, parent, false) as EventItemView
            ItemViewType.TYPE_OTHER -> layoutInflater.inflate(R.layout.item_schedule_event_other, parent, false) as EventItemView
        }
        return EventViewHolder(itemView)
    }

    override fun onBindViewHolder(holder: EventViewHolder, position: Int) {
        holder.updateWith(getItem(position), showRoom, showFavorites, eventClickListener)
    }

    @Deprecated(
        message = "Use updateWith() instead",
        replaceWith = ReplaceWith("updateWith(list, showRoom, eventClickListener)"),
        level = DeprecationLevel.ERROR
    )
    override fun submitList(list: MutableList<Event>?) {
        throw UnsupportedOperationException("Use updateWith() instead")
    }
}

internal class EventViewHolder(itemView: EventItemView) : RecyclerView.ViewHolder(itemView) {

    fun updateWith(event: Event, showRoom: Boolean, showFavorites: Boolean, listener: OnEventClickListener?) {
        (itemView as EventItemView).updateWith(event, showRoom = showRoom, showFavorite = showFavorites)
        itemView.setOnClickListener { listener?.invoke(event) }
    }
}

private object DiffCallback : DiffUtil.ItemCallback<Event>() {

    override fun areItemsTheSame(oldItem: Event, newItem: Event): Boolean {
        return oldItem.numericId == newItem.numericId
    }

    override fun areContentsTheSame(oldItem: Event, newItem: Event): Boolean {
        return oldItem == newItem
    }
}

private enum class ItemViewType {
    TYPE_TALK,
    TYPE_OTHER
}

typealias OnEventClickListener = (Event) -> Unit
